mdn/developer-portal
# 1669: Add rate limiting

This changeset addresses the need to rate limit Wagtail pages which support search and the custom site-search view. 

This is needed because pages which process a search input (Posts, Events and Site Search) can accept a huge variety of  search args via GET param, and so these pages will likely **not** already be cached in the CDN (unless someone has searched for the same term since the CDN was last invalidated). Rate limiting will protect us from unnecessary load due to bad actors / misbehaving bots. 

The rate limiting is carried out by adding custom middleware that uses a core (public) helper function from `django-ratelimit`, which is what is also used in Kuma. It rate-limits ALL pages, currently to 25 requests per IP per minute. (Details of the decision are below)


----


While regular Django function- or class-based views can be easily rate-limited via a decorator, it is not possible to do the same in a Wagtail project, because most pages in the site are handled by Wagtail's own routing/resolving logic, which is not easy to decorate. Hence the choice to do it with middleware which can "get in front of the" page serving logic and return early if need be.

This decision also means the _entire_ site is rate-limited - currently at a limit of 25 requests per IP per minute - rather than only the views that the CDN will not be cacheing 

When an IP is rate-limited, we send the user to a custom page

<img width="1679" alt="Screenshot 2020-07-06 at 23 52 38" src="https://user-images.githubusercontent.com/101457/86846441-c6c11900-c0a2-11ea-9f93-f185f8269b01.png">

(While in the neighbourhood, I also made a custom 403 page)

<img width="1255" alt="Screenshot 2020-07-07 at 00 19 17" src="https://user-images.githubusercontent.com/101457/86846480-d3457180-c0a2-11ea-88b4-a80ba74388f4.png">



---

**Why are we rate-limiting ALL views and not just those that the CDN won't be cacheing? Here's the logic:**

* Most production traffic spikes will be handled by the CDN. The most likely pages
to be CDN cache misses are search ones, because the GET params can contain anything
* Rate-limiting Search is the real goal here, but we have challenges to overcome.
* Search appears on three pages - one is a vanilla Django view `search:site-search`
two are Wagtail pages which are served by `wagtail.core.views.serve`, and those
pages handle GET-based search queries themselves -- they don't hand off to the
`search:site-search` view
* We can't decorate the Wagtail `serve` view with `django-ratelimit`, because it is buried inside Wagtail, but we can use
middleware to do perform rate-limiting checks, using `django-ratelimit`'s core helpers
* However, `django-ratelimit`'s API/helpers need a `HTTPRequest` object, because they
rate-limit based on a view func, _not_ a path from the URL, so we can't easily target
specific Wagtail-served pages for rate limiting.

So: given that non-search pages will be cached by the CDN, we rate-limit ALL the
pages served by the site. In the wild, once the CDN is warm, the majority of
uncached requests will be search ones, so we'll effectively be only rate-limiting
them.

----
(Related issue #1669)

## Key changes:

* Ensure Site Search page is out of scope for spiders - a one-line change that brings portal in line with kuma
* Add django-ratelimter to project + HTTP 429 template, to show to throttled users + add HTTP 403 template while in the neighbourhood
* Add middleware to enforce rate limiting
* Ensure rate-limiting is disabled for health check tests
* Add tests for rate-limiting middleware, to confirm it's using django-ratelimit's internals effectively
* Add in middleware to ensure X-Forwarded-For header is processed appropriately so rate-limiting works on client IP, not CDN or ELB IP. Based on equivalent middleware in Kuma.
* 1669: Add test for the rate_limited HTTP 429 view 

## How to test

This one's less straightforward to test, so I would recommend code review as the main thing, but as far as a check goes, with the knowledge that requests are currently limited to 25/min per IP address on stage (where the code is deployed)

1) Load any page on the non-CDN-fronted staging site repeatedly, 30 times. Your 26th attempt will trigger the rate limiting, for a minute. eg: https://developer-portal.stage.mdn.mozit.cloud/search/?search=test
  * When you get the rate-limited page, use devtools to inspect the response headers and confirm you see the `Cache-Control` set as `max-age=0, no-cache, no-store, must-revalidate` and the HTTP response code as 429

2) Load that same page via the CDN-fronted staging site: https://developer-portal-cdn.stage.mdn.mozit.cloud/search/?search=test and you will find you can hit it again and again, way more than 26 times and not get rate limited, because the CDN is serving the page from its cache. (Responses will be HTTP 200 always and Devtools' network pane will show it as a `Hit from Cloudfront`
